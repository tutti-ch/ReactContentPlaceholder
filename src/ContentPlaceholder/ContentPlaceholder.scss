@import 'style_guide';

@keyframes placeHolderShimmer {
  0% {
    background-position: -468px 0
  }
  100% {
    background-position: 468px 0
  }
}

.animatedBg {
  animation-duration: 1s;
  animation-fill-mode: forwards;
  animation-iteration-count: infinite;
  animation-name: placeHolderShimmer;
  animation-timing-function: linear;
  background: $greyscale-90 linear-gradient(to right, $greyscale-90 8%, $greyscale-95 18%, $greyscale-90 33%);
  background-size: 800px 104px;
  position: relative;
  height: 0.4rem;
  margin-bottom: $size-xxs;
  width: 60%;

  // Every fourth, starting at the first
  &:nth-child(4n + 1) {
    width: 80%;
  }

  // Every even
  &:nth-child(even) {
    width: 90%;
  }
}

.paragraph {
  margin-bottom: $size-m;

  &:last-child {
    margin-bottom: 0;
  }
}

.imageText {
  margin-bottom: $size-m;
  $image-size: 3rem;
  display: flex;
  align-items: center;
  justify-content: center;

  &.vertical {
    flex-direction: column;

    .contentLeft {
      margin-right: 0;
      flex: 1 1 auto;
    }

    .contentRight {
      width: 100%;
      margin-top: $size-xs;

      .animatedBg {
        margin-left: auto;
        margin-right: auto;
      }
    }
  }

  .contentLeft {
    flex: 0 0 auto;
    width: $image-size;
    margin-right: $size-l;
  }

  .contentRight {
    flex: 1 1 auto;

    .animatedBg:first-child {
      height: 0.7rem;
    }
  }

  .animatedBg.image {
    height: $image-size;
    width: $image-size;
    border-radius: 50%;
  }
}

.card {
  border: 1px solid $border-default;
  padding: $size-m;
  display: flex;

  .cardImage {
    height: 5rem;
    flex: 0 0 auto;
    width: 30%;
    margin-right: $size-m;
  }

  .paragraph {
    flex: 1 1 auto;
  }

  &.vertical {
    flex-direction: column;
    width: 100%;

    .cardImage {
      flex: 1 1 auto;
      width: 100%;
    }

    .paragraph {
      width: 100%;
      margin-top: $size-xs;
    }
  }
}
