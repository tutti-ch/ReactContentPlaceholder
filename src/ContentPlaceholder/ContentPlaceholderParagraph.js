import React, { Component } from 'react'
import PropTypes from 'prop-types'

import classes from './ContentPlaceholder.scss'

/**
 * Paragraph component for the ContentPlaceholder.
 *
 * Simply use ContentPlaceholder.Paragraph to access this component.
 */
export default class ContentPlaceholderParagraph extends Component {
  static propTypes = {
    numberOfLines: PropTypes.number
  }

  static defaultProps = {
    numberOfLines: 3
  }

  /**
   * Render the component with specified number of lines. This component
   * expects a non-zero positive integer. By default it will render 3 lines.
   *
   * @return {XML}
   */
  render () {
    const { numberOfLines } = this.props

    return (
      <div className={classes.paragraph}>
        {
          [...new Array(numberOfLines).keys()].map((_, index) => (
            <div key={index} className={`${classes.animatedBg}`}></div>
          ))
        }
      </div>
    )
  }
}
