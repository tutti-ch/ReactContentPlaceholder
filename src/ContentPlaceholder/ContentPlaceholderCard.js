import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import classes from './ContentPlaceholder.scss'
import ContentPlaceholderParagraph from './ContentPlaceholderParagraph'

export default class ContentPlaceholderCard extends Component {
  static propTypes = {
    align: PropTypes.oneOf(['horizontal', 'vertical']),
    numberOfLines: PropTypes.number
  }

  static defaultProps = {
    align: 'horizontal',
    numberOfLines: null
  }

  render () {
    const { align } = this.props
    let { numberOfLines } = this.props

    if (numberOfLines === null) {
      numberOfLines = align === 'horizontal' ? 6 : 2
    }

    return (
      <div className={classNames(classes.card, classes[align])}>
        <div className={`${classes.animatedBg} ${classes.cardImage}`}></div>
        <ContentPlaceholderParagraph numberOfLines={numberOfLines} />
      </div>
    )
  }
}
