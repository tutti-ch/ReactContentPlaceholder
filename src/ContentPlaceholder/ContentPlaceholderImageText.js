import React, { Component } from 'react'
import PropTypes from 'prop-types'

import ContentPlaceholderParagraph from './ContentPlaceholderParagraph'
import classes from './ContentPlaceholder.scss'

/**
 * This component will create an image with text on.
 *
 * Simply use ContentPlaceholder.ImageText to access this component.
 */
export default class ContentPlaceholderImageText extends Component {
  static propTypes = {
    align: PropTypes.oneOf(['horizontal', 'vertical']),
    numberOfLines: PropTypes.number
  }

  static defaultProps = {
    align: 'horizontal',
    numberOfLines: 2
  }

  /**
   * Depending on the withImage property, renders the component with or without image.
   *
   * @return {ReactElement}
   */
  render () {
    const { align } = this.props

    return (
      <div className={`${classes.imageText} ${classes[align]}`}>
        <div className={classes.contentLeft}>
          <div className={`${classes.animatedBg} ${classes.image}`}></div>
        </div>
        <div className={classes.contentRight}>
          <ContentPlaceholderParagraph numberOfLines={2} />
        </div>
      </div>
    )
  }
}
