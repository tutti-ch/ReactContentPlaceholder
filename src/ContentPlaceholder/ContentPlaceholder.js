import React, { Component } from 'react'
import PropTypes from 'prop-types'

import ContentPlaceholderParagraph from './ContentPlaceholderParagraph'
import ContentPlaceholderImageText from './ContentPlaceholderImageText'
import ContentPlaceholderCard from './ContentPlaceholderCard'

export default class ContentPlaceholder extends Component {
  static propTypes = {
    className: PropTypes.oneOfType([PropTypes.string, PropTypes.object, PropTypes.array]),
    children: PropTypes.node
  }

  static ImageText = ContentPlaceholderImageText
  static Paragraph = ContentPlaceholderParagraph
  static Card = ContentPlaceholderCard

  /**
   * This will render the component.
   *
   * @example
   * <ContentPlaceholder className='my-class'>
   *   <div>
   *     <ContentPlaceholder.Header withImage />
   *     <ContentPlaceholder.Paragraph numberOfLines={3} />
   *     <ContentPlaceholder.Paragraph numberOfLines={1} />
   *   </div>
   * </ContentPlaceholder>
   *
   * @return {ReactElement}
   */
  render () {
    const { className, children } = this.props

    return (
      <div className={className}>
        {children}
      </div>
    )
  }
}
