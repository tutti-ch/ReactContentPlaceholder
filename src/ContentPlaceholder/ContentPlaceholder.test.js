/* global describe, expect, test */
import React from 'react'
import ContentPlaceholder from './index'
import renderer from 'react-test-renderer'

describe('(Component) ContentPlaceholder', () => {
  test('should render', () => {
    const withCard = renderer.create(
      <ContentPlaceholder>
        <ContentPlaceholder.Card align="horizontal" numberOfLines={3} />
      </ContentPlaceholder>
    )

    const withImageText = renderer.create(
      <ContentPlaceholder>
        <ContentPlaceholder.ImageText align="vertical" numberOfLines={2} />
      </ContentPlaceholder>
    )

    const moreComplicated = renderer.create(
      <ContentPlaceholder>
        <ContentPlaceholder.ImageText align="horizontal" />
        <ContentPlaceholder.Card align="vertical" />
        <ContentPlaceholder.Paragraph numberOfLines={4} />
        <ContentPlaceholder.Paragraph numberOfLines={2} />
      </ContentPlaceholder>
    )

    expect(withCard).toMatchSnapshot()
    expect(withImageText).toMatchSnapshot()
    expect(moreComplicated).toMatchSnapshot()
  })
})
