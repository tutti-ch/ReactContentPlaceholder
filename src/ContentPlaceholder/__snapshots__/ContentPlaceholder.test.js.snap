// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) ContentPlaceholder should render 1`] = `
<div
  className={undefined}
>
  <div
    className=""
  >
    <div
      className="undefined undefined"
    />
    <div
      className={undefined}
    >
      <div
        className="undefined"
      />
      <div
        className="undefined"
      />
      <div
        className="undefined"
      />
    </div>
  </div>
</div>
`;

exports[`(Component) ContentPlaceholder should render 2`] = `
<div
  className={undefined}
>
  <div
    className="undefined undefined"
  >
    <div
      className={undefined}
    >
      <div
        className="undefined undefined"
      />
    </div>
    <div
      className={undefined}
    >
      <div
        className={undefined}
      >
        <div
          className="undefined"
        />
        <div
          className="undefined"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`(Component) ContentPlaceholder should render 3`] = `
<div
  className={undefined}
>
  <div
    className="undefined undefined"
  >
    <div
      className={undefined}
    >
      <div
        className="undefined undefined"
      />
    </div>
    <div
      className={undefined}
    >
      <div
        className={undefined}
      >
        <div
          className="undefined"
        />
        <div
          className="undefined"
        />
      </div>
    </div>
  </div>
  <div
    className=""
  >
    <div
      className="undefined undefined"
    />
    <div
      className={undefined}
    >
      <div
        className="undefined"
      />
      <div
        className="undefined"
      />
    </div>
  </div>
  <div
    className={undefined}
  >
    <div
      className="undefined"
    />
    <div
      className="undefined"
    />
    <div
      className="undefined"
    />
    <div
      className="undefined"
    />
  </div>
  <div
    className={undefined}
  >
    <div
      className="undefined"
    />
    <div
      className="undefined"
    />
  </div>
</div>
`;
